packages/realm/src/Logger.ts

```diff
@@ -0,0 +1,91 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { assert, binding } from "./internal";
+
+export type LogLevel = "all" | "trace" | "debug" | "detail" | "info" | "warn" | "error" | "fatal" | "off";
+
+export enum NumericLogLevel {
+  All = 0,
+  Trace = 1,
+  Debug = 2,
+  Detail = 3,
+  Info = 4,
+  Warn = 5,
+  Error = 6,
+  Fatal = 7,
+  Off = 8,
+}
+
+export type Logger = (level: NumericLogLevel, message: string) => void;
+
+export type LoggerCallback = (level: LogLevel, message: string) => void;
+
+/** @internal */
+export function toBindingLoggerLevel(arg: LogLevel): binding.LoggerLevel {
+  const bindingLogLevel = inverseTranslationTable[arg];
+  assert(bindingLogLevel !== undefined, `Unexpected log level: ${arg}`);
+  return bindingLogLevel;
+}
+
+/** @internal */
+export function fromBindingLoggerLevelToNumericLogLevel(arg: binding.LoggerLevel): NumericLogLevel {
+  // For now, these map 1-to-1
+  return arg as unknown as NumericLogLevel;
+}
+
+const translationTable: Record<binding.LoggerLevel, LogLevel> = {
+  [binding.LoggerLevel.All]: "all",
+  [binding.LoggerLevel.Trace]: "trace",
+  [binding.LoggerLevel.Debug]: "debug",
+  [binding.LoggerLevel.Detail]: "detail",
+  [binding.LoggerLevel.Info]: "info",
+  [binding.LoggerLevel.Warn]: "warn",
+  [binding.LoggerLevel.Error]: "error",
+  [binding.LoggerLevel.Fatal]: "fatal",
+  [binding.LoggerLevel.Off]: "off",
+};
+
+const inverseTranslationTable: Record<LogLevel, binding.LoggerLevel> = Object.fromEntries(
+  Object.entries(translationTable).map(([key, val]) => [val, Number(key)]),
+) as unknown as Record<LogLevel, binding.LoggerLevel>;
```

Comment:
I'd like to iterate on this with you. I think it'd be possible to use the type parameters on the `fromEntries`, `entries` and `map` to achieve this and avoid `as unknown`.

Reply:
Sure! Let me know when you want to go over it together

Reply:
I actually discovered that we don't need the `unknown` 😁 
Before I didn't have the conversion to `number`, and that was the issue